File: 1269-Num-Ways/1269.py
class Solution:
  def numWays(self, steps: int, arrLen: int) -> int:
    mod = 10 ** 9 + 7
    maxColumn = min(arrLen - 1, steps)

    dp = [0] * (maxColumn + 1)
    dp[0] = 1

    for i in range(1, steps + 1):
      dpNext = [0] * (maxColumn + 1)
      for j in range(0, maxColumn + 1):
        dpNext[j] = dp[j]
        if j - 1 >= 0:
          dpNext[j] = (dpNext[j] + dp[j - 1]) % mod
        if j + 1 <= maxColumn:
          dpNext[j] = (dpNext[j] + dp[j - 1]) % mod
      dp = dpNext

    return dp[0]

File: 1269-Num-Ways/1269.java
class Solution {
  public int numWays(int steps, int arrLen) {
    final int MODULO = 1000000007;
    int maxColumn = Math.min(arrLen - 1, steps);
    int[] dp = new int[maxColumn + 1];
    dp[0] = 1;
    for (int i = 1; i <= steps; i++) {
      int[] dpNext = new int[maxColumn + 1];
      for (int j = 0; j <= maxColumn; j++) {
        dpNext[j] = dp[j];
        if (j - 1 >= 0) {
          dpNext[j] = (dpNext[j] + dp[j - 1]) % MODULO;
        }
        if (j + 1 <= maxColumn) {
          dpNext[j] = (dpNext[j] + dp[j + 1]) % MODULO;
        }
      }
      dp = dpNext;
    }
    return dp[0];
  }
}